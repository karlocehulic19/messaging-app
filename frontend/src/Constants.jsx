const prod = {
  url: {
    BACKEND_URL: import.meta.env.VITE_PROD_BACKEND_URL,
  },
};

const dev = {
  url: {
    BACKEND_URL: import.meta.env.VITE_DEV_BACKEND_URL,
  },
};

export const config = import.meta.env.MODE === "production" ? prod : dev;
