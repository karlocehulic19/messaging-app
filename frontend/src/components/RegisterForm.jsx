import { useNavigate } from "react-router-dom";
import RegistrationValidator from "../utils/RegistrationValidator";
import useValidator from "../hooks/useValidator";

class InputShortcut {
  constructor(name, password = false) {
    this.props = {
      name,
      id: name,
      type: "text",
      ["data-testid"]: !password ? undefined : "pw-input",
    };
  }
  addProp(prop, value) {
    this.props[prop] = value;

    return this;
  }
  getProps() {
    return this.props;
  }
  getField() {
    return this.props.name;
  }
  getPlaceholder() {
    if (!this.props.placeholder)
      throw new Error(`${this.getField()} doesn't have placeholder`);
    return this.props.placeholder;
  }
}

const defaultInputs = [
  new InputShortcut("username").addProp("placeholder", "Username"),
  new InputShortcut("email")
    .addProp("placeholder", "Email")
    .addProp("type", "email"),
  new InputShortcut("firstName").addProp("placeholder", "First Name"),
  new InputShortcut("lastName").addProp("placeholder", "Last Name"),
  new InputShortcut("password", true)
    .addProp("placeholder", "Password")
    .addProp("aria-label", "Password input")
    .addProp("type", "password"),
  new InputShortcut("passwordConf", true)
    .addProp("placeholder", "Confirm Password")
    .addProp("aria-label", "Password Confirmation input")
    .addProp("type", "password"),
];

function RegisterForm() {
  const navigate = useNavigate();
  const { formData, changeFormData, validationErrors } = useValidator(
    RegistrationValidator
  );

  function handleSubmit(e) {
    e.preventDefault();
    const hasErrors = !!Object.keys(validationErrors).length;
    if (!hasErrors) navigate("/login");
  }

  function handleInputChange(e) {
    changeFormData(e.target.id, e.target.value);
  }

  return (
    <form noValidate onSubmit={handleSubmit} aria-label="Login form">
      {defaultInputs.map((shortcut, index) => {
        return (
          <label
            data-testid={"input-label"}
            key={index}
            htmlFor={shortcut.getField()}
          >
            {validationErrors[shortcut.getField()] && (
              <span
                aria-label={`${shortcut.getPlaceholder()} validation error`}
                data-testid={"validation-msg"}
              >
                {validationErrors[shortcut.getField()]}
              </span>
            )}
            <input
              value={
                formData[shortcut.getField()]
                  ? formData[shortcut.getField()]
                  : ""
              }
              onChange={handleInputChange}
              {...shortcut.getProps()}
            ></input>
          </label>
        );
      })}
      <button type="submit">Register</button>
    </form>
  );
}

export default RegisterForm;
