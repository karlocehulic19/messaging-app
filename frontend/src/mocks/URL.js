global.URL.createObjectURL = vi.fn(() => "mockedFilePath");
global.URL.revokeObjectURL = vi.fn(() => null);
