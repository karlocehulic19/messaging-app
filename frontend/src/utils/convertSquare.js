import { Jimp } from "jimp";

export default async function convertSquare(
  pictureBuffer,
  pictureType,
  size = 200
) {
  const picture = await Jimp.fromBuffer(pictureBuffer);
  if (size > picture.height || size > picture.width)
    throw new Error("Too big crop size");
  picture.cover({ w: size, h: size });
  return await picture.getBase64(pictureType);
}
