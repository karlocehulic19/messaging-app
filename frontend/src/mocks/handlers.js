import { http, HttpResponse } from "msw";
import { config } from "../Constants";

export const handlers = [
  http.post(`${config.url.BACKEND_URL}/login`, async ({ request }) => {
    const body = await request.json();
    if (!body.username || !body.password) {
      return new HttpResponse("Missing credentials", { status: 401 });
    }

    if (body.username != "someUsername" || body.password != "somePassword") {
      return HttpResponse.json(
        { messages: ["Username or password is incorrect"] },
        { status: 401 }
      );
    }
    return HttpResponse.json({ token: "randomJWTtoken" }, { status: 200 });
  }),
];
