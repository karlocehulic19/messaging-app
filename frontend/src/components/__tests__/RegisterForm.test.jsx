import { cleanup, render, screen, act } from "@testing-library/react";
import RegisterForm from "../RegisterForm";
import { expect, describe, it, vi, afterEach } from "vitest";
import userEvent from "@testing-library/user-event";
import "../../mocks/URL";
import { server } from "../../mocks/node";
import { config } from "../../Constants";
import { HttpResponse, http } from "msw";
import { readFileSync } from "node:fs";
import path from "node:path";
import convertSquare from "../../utils/convertSquare";

const mockedNavigate = vi.fn();
server.listen();

const registerHandler = (resolver) =>
  http.post(`${config.url.BACKEND_URL}/register`, resolver);

const resolver500 = () =>
  HttpResponse.json({ error: "Internal Server Error" }, { status: 500 });

const resolver422 = () =>
  HttpResponse.json(
    {
      message: ["First validation message", "Second validation message"],
    },
    { status: 422 }
  );
vi.mock("react-router-dom", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useNavigate: () => mockedNavigate,
  };
});

afterEach(() => {
  server.resetHandlers();
  vi.clearAllMocks();
});

const setup = async (userEventOptions = {}) => {
  const user = userEvent.setup(userEventOptions);
  render(<RegisterForm />);

  await user.click(screen.getByPlaceholderText("Username"));
  await user.keyboard("test");

  await user.click(screen.getByPlaceholderText("First Name"));
  await user.keyboard("Karlo");

  await user.click(screen.getByPlaceholderText("Last Name"));
  await user.keyboard("ÄŒehuliÄ‡");

  await user.click(screen.getByPlaceholderText("Email"));
  await user.keyboard("karlocehlic@gmail.com");

  await user.click(screen.getByPlaceholderText("Password"));
  await user.keyboard("V4l1dP4ssw@rd");

  await user.click(screen.getByPlaceholderText("Confirm Password"));
  await user.keyboard("V4l1dP4ssw@rd");

  return { user };
};

const setupNearlyCorrect = async () => {
  const user = userEvent.setup();
  render(<RegisterForm />);

  await user.click(screen.getByPlaceholderText("Email"));
  await user.keyboard("karlocehlic@gmail");

  await user.click(screen.getByPlaceholderText("Password"));
  await user.keyboard("V4l1dP4ssw@rd");

  await user.click(screen.getByPlaceholderText("Confirm Password"));
  await user.keyboard("V4l1dP4ssw@r");

  return { user };
};

it("renders form with right attributes", () => {
  render(<RegisterForm />);

  expect(screen.getByRole("form").attributes).toMatchSnapshot();
});

it("renders labels with right attribute", () => {
  render(<RegisterForm />);
  expect(
    screen.getAllByTestId("input-label").map((ele) => ele.attributes)
  ).toMatchSnapshot();
});

it("renders empty inputs correctly(non-password)", () => {
  render(<RegisterForm />);
  expect(screen.getAllByRole("textbox")).toMatchSnapshot();
});

it("renders empty inputs correctly(password)", () => {
  render(<RegisterForm />);
  expect(screen.getAllByTestId("pw-input")).toMatchSnapshot();
});

describe("validation logic", () => {
  it("doesn't render validation errors after first render", () => {
    render(<RegisterForm />);

    expect(screen.queryAllByTestId("validation-msg")).toEqual([]);
  });

  it("renders validation errors after interacting with input", async () => {
    render(<RegisterForm />);
    const user = userEvent.setup();

    await user.click(screen.getByPlaceholderText("Username"));

    await user.keyboard("a");

    expect(screen.getAllByTestId("validation-msg")).toMatchSnapshot();
  });

  it("renders validation errors after interacting with any input", async () => {
    render(<RegisterForm />);
    const user = userEvent.setup();

    await user.click(screen.getByPlaceholderText("Password"));
    await user.keyboard("a");

    expect(screen.getAllByTestId("validation-msg")).toMatchSnapshot();
  });

  it("doesn't redirect after only clicking on register", async () => {
    render(<RegisterForm />);
    const user = userEvent.setup();

    await user.click(screen.getByRole("button"));

    expect(mockedNavigate).not.toBeCalled();
  });

  it("redirects after right validation", async () => {
    const { user } = await setup();

    await user.click(screen.getByRole("button"));
    expect(mockedNavigate).toBeCalledTimes(1);
    expect(mockedNavigate).toBeCalledWith("/login");
  });

  it("doesn't redirect after wrong validation", async () => {
    const { user } = await setup();

    await user.click(screen.getByPlaceholderText("Confirm Password"));
    await user.keyboard("1");
    await user.click(screen.getByRole("button"));

    expect(mockedNavigate).toBeCalledTimes(0);
  });

  it("renders one invalid input correctly(passConfirmation)", async () => {
    const { user } = await setup();

    await user.click(screen.getByPlaceholderText("Confirm Password"));
    await user.keyboard("1");
    await user.click(screen.getByRole("button"));

    expect(screen.getByTestId("validation-msg")).toMatchSnapshot();
  });

  it("renders many invalid inputs correctly(all of them in nearly correct setup)", async () => {
    const { user } = await setupNearlyCorrect();
    await user.click(screen.getByRole("button"));

    expect(screen.getAllByTestId("validation-msg")).toMatchSnapshot();
  });

  it("makes register button disabled with wrong validation", async () => {
    render(<RegisterForm />);

    expect(screen.getByRole("button")).toBeDisabled();

    cleanup();

    await setupNearlyCorrect();
    expect(screen.getByRole("button")).toBeDisabled();
  });
});

describe("fetching logic", () => {
  it("redirects after ok response", async () => {
    const { user } = await setup();
    await user.click(screen.getByRole("button"));

    expect(mockedNavigate).toHaveBeenCalledOnce();
  });

  it("doesn't redirect after bad responses", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);

    server.use(registerHandler(resolver500));
    const { user: user1 } = await setup();

    await user1.click(screen.getByRole("button"));

    expect(mockedNavigate).not.toHaveBeenCalled();

    cleanup();

    server.use(registerHandler(resolver422));
    const { user: user2 } = await setup();
    await user2.click(screen.getByRole("button"));

    expect(mockedNavigate).not.toHaveBeenCalled();
  });

  it("doesn't displays an error popup on meant requests", async () => {
    const { user: user1 } = await setup();
    await user1.click(screen.getByRole("button"));

    expect(screen.queryByLabelText("Error message")).not.toBeInTheDocument();

    cleanup();

    const { user: user2 } = await setup();
    server.use(registerHandler(resolver422));
    await user2.click(screen.getByRole("button"));

    expect(screen.queryByLabelText("Error message")).not.toBeInTheDocument();
  });

  it("displays an error popup on failed requests", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);
    const { user } = await setup();
    server.use(registerHandler(resolver500));
    await user.click(screen.getByRole("button"));

    expect(screen.getByLabelText("Error message")).toMatchSnapshot();
  });

  it("displays an error popup on failed network request", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);
    server.use(registerHandler(() => HttpResponse.error()));
    const { user } = await setup();

    await user.click(screen.getByRole("button"));

    expect(screen.getByLabelText("Error message")).toMatchSnapshot();
  });

  it("displays validation messages on 422", async () => {
    server.use(registerHandler(resolver422));
    const { user } = await setup();

    await user.click(screen.getByRole("button"));

    expect(screen.getByText("First validation message")).toMatchSnapshot();
    expect(screen.getByText("Second validation message")).toMatchSnapshot();
  });

  it("displays validation messages anything other than 400", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);
    const { user: user1 } = await setup();

    await user1.click(screen.getByRole("button"));

    expect(screen.queryByText("First validation message")).toBeNull();
    expect(screen.queryByText("Second validation message")).toBeNull();

    cleanup();

    server.use(registerHandler(resolver500));
    const { user: user2 } = await setup();

    await user2.click(screen.getByRole("button"));

    expect(screen.queryByText("First validation message")).toBeNull();
    expect(screen.queryByText("Second validation message")).toBeNull();
  });

  it("disables register button until request is received", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);
    let requestResolver;

    server.use(
      registerHandler(
        async () =>
          await new Promise((resolve) => {
            requestResolver = () => resolve(HttpResponse.error());
          })
      )
    );

    const { user } = await setup();
    await user.click(screen.getByRole("button"));

    expect(screen.getByRole("button")).toBeDisabled();

    await act(async () => {
      requestResolver();
    });

    expect(screen.getByRole("button")).toBeEnabled();
  });

  it("sets register button text to loading request is received", async () => {
    vi.spyOn(console, "error").mockImplementation(() => undefined);
    let requestResolver;

    server.use(
      registerHandler(
        async () =>
          await new Promise((resolve) => {
            requestResolver = () => resolve(HttpResponse.error());
          })
      )
    );

    const { user } = await setup();
    await user.click(screen.getByRole("button"));

    expect(screen.getByRole("button").textContent).toBe("Loading...");

    await act(async () => {
      requestResolver();
    });

    expect(screen.getByRole("button").textContent).toBe("Register");
  });

  describe("profile picture", () => {
    it("calls api with formatted base64 of included picture", async () => {
      const { user } = await setup();
      const fetchSpy = vi.spyOn(global, "fetch");
      const jpgBuffer = readFileSync(
        // eslint-disable-next-line no-undef
        path.resolve(__dirname, "../../tests/assets/jpg-file.jpg")
      );
      const jpgTestFile = new File([jpgBuffer], "test.jpg", {
        type: "image/jpeg",
      });

      await user.upload(screen.getByTestId("picture-input"), jpgTestFile);
      await user.click(screen.getByLabelText("Submit Button"));

      expect(fetchSpy).toBeCalledTimes(1);
      expect(fetchSpy).toBeCalledWith(`${config.url.BACKEND_URL}/register`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          username: "test",
          firstName: "Karlo",
          lastName: "ÄŒehuliÄ‡",
          email: "karlocehlic@gmail.com",
          password: "V4l1dP4ssw@rd",
          pictureBase64: await convertSquare(jpgBuffer, "image/jpeg"),
        }),
      });
    });

    it("doesn't call api with formatted base64 when pic not selected", async () => {
      const { user } = await setup();
      const fetchSpy = vi.spyOn(global, "fetch");
      await user.click(screen.getByLabelText("Submit Button"));

      expect(fetchSpy).toBeCalledTimes(1);
      expect(fetchSpy).toBeCalledWith(`${config.url.BACKEND_URL}/register`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          username: "test",
          firstName: "Karlo",
          lastName: "ÄŒehuliÄ‡",
          email: "karlocehlic@gmail.com",
          password: "V4l1dP4ssw@rd",
        }),
      });
    });

    it("doesn't call api with formatted base64 when pic is removed", async () => {
      const { user } = await setup();
      const fetchSpy = vi.spyOn(global, "fetch");
      const jpgBuffer = readFileSync(
        // eslint-disable-next-line no-undef
        path.resolve(__dirname, "../../tests/assets/jpg-file.jpg")
      );
      const jpgTestFile = new File([jpgBuffer], "test.jpg", {
        type: "image/jpeg",
      });

      await user.upload(screen.getByTestId("picture-input"), jpgTestFile);

      await user.click(screen.getByLabelText("Remove Picture"));

      await user.click(screen.getByLabelText("Submit Button"));

      expect(fetchSpy).toBeCalledTimes(1);
      expect(fetchSpy).toBeCalledWith(`${config.url.BACKEND_URL}/register`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          username: "test",
          firstName: "Karlo",
          lastName: "ÄŒehuliÄ‡",
          email: "karlocehlic@gmail.com",
          password: "V4l1dP4ssw@rd",
        }),
      });
    });
  });
});
