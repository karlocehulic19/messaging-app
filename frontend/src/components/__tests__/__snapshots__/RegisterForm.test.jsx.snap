// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetching logic > displays an error popup on failed network request 1`] = `
<div
  aria-label="Error message"
  class="error-popup"
  role="alert"
>
  <p>
    An error occurred. Please try again.
  </p>
</div>
`;

exports[`fetching logic > displays an error popup on failed requests 1`] = `
<div
  aria-label="Error message"
  class="error-popup"
  role="alert"
>
  <p>
    An error occurred. Please try again.
  </p>
</div>
`;

exports[`fetching logic > displays validation messages on 422 1`] = `
<p>
  First validation message
</p>
`;

exports[`fetching logic > displays validation messages on 422 2`] = `
<p>
  Second validation message
</p>
`;

exports[`renders empty inputs correctly(non-password) 1`] = `
[
  <input
    id="username"
    name="username"
    placeholder="Username"
    type="text"
    value=""
  />,
  <input
    id="email"
    name="email"
    placeholder="Email"
    type="email"
    value=""
  />,
  <input
    id="firstName"
    name="firstName"
    placeholder="First Name"
    type="text"
    value=""
  />,
  <input
    id="lastName"
    name="lastName"
    placeholder="Last Name"
    type="text"
    value=""
  />,
]
`;

exports[`renders empty inputs correctly(password) 1`] = `
[
  <input
    aria-label="Password input"
    data-testid="pw-input"
    id="password"
    name="password"
    placeholder="Password"
    type="password"
    value=""
  />,
  <input
    aria-label="Password Confirmation input"
    data-testid="pw-input"
    id="passwordConf"
    name="passwordConf"
    placeholder="Confirm Password"
    type="password"
    value=""
  />,
]
`;

exports[`renders form with right attributes 1`] = `
NamedNodeMap {
  "aria-label": "Login form",
  "class": "_register-form_9e6f01",
  "novalidate": "",
}
`;

exports[`renders labels with right attribute 1`] = `
[
  NamedNodeMap {
    "data-testid": "input-label",
    "for": "username",
  },
  NamedNodeMap {
    "data-testid": "input-label",
    "for": "email",
  },
  NamedNodeMap {
    "data-testid": "input-label",
    "for": "firstName",
  },
  NamedNodeMap {
    "data-testid": "input-label",
    "for": "lastName",
  },
  NamedNodeMap {
    "data-testid": "input-label",
    "for": "password",
  },
  NamedNodeMap {
    "data-testid": "input-label",
    "for": "passwordConf",
  },
]
`;

exports[`validation logic > renders many invalid inputs correctly(all of them in nearly correct setup) 1`] = `
[
  <span
    aria-label="Username validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Username field is required
  </span>,
  <span
    aria-label="Email validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Value provided must be an valid email address
  </span>,
  <span
    aria-label="First Name validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    First name is required
  </span>,
  <span
    aria-label="Last Name validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Last name is required
  </span>,
  <span
    aria-label="Confirm Password validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Password confirmation must be same as password
  </span>,
]
`;

exports[`validation logic > renders one invalid input correctly(passConfirmation) 1`] = `
<span
  aria-label="Confirm Password validation error"
  class="_client-validation-error_4fc6c6"
  data-testid="validation-msg"
>
  Password confirmation must be same as password
</span>
`;

exports[`validation logic > renders validation errors after interacting with any input 1`] = `
[
  <span
    aria-label="Username validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Username field is required
  </span>,
  <span
    aria-label="Email validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Email field is required
  </span>,
  <span
    aria-label="First Name validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    First name is required
  </span>,
  <span
    aria-label="Last Name validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Last name is required
  </span>,
  <span
    aria-label="Password validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Password must contain at least one uppercase letter
  </span>,
  <span
    aria-label="Confirm Password validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Password confirmation is required
  </span>,
]
`;

exports[`validation logic > renders validation errors after interacting with input 1`] = `
[
  <span
    aria-label="Email validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Email field is required
  </span>,
  <span
    aria-label="First Name validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    First name is required
  </span>,
  <span
    aria-label="Last Name validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Last name is required
  </span>,
  <span
    aria-label="Password validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Password field is required
  </span>,
  <span
    aria-label="Confirm Password validation error"
    class="_client-validation-error_4fc6c6"
    data-testid="validation-msg"
  >
    Password confirmation is required
  </span>,
]
`;
